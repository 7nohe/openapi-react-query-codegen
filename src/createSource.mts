import { join } from "node:path";
import { Project } from "ts-morph";
import ts from "typescript";
import { OpenApiRqFiles } from "./constants.mjs";
import { createExports } from "./createExports.mjs";
import { createImports } from "./createImports.mjs";
import { getServices } from "./service.mjs";

const createSourceFile = async (outputPath: string) => {
  const project = new Project({
    // Optionally specify compiler options, tsconfig.json, in-memory file system, and more here.
    // If you initialize with a tsconfig.json, then it will automatically populate the project
    // with the associated source files.
    // Read more: https://ts-morph.com/setup/
    skipAddingFilesFromTsConfig: true,
  });

  const sourceFiles = join(process.cwd(), outputPath);
  project.addSourceFilesAtPaths(`${sourceFiles}/**/*`);

  const service = await getServices(project);

  const imports = createImports({
    project,
  });

  const exports = createExports(service);

  const commonSource = ts.factory.createSourceFile(
    [...imports, ...exports.allCommon],
    ts.factory.createToken(ts.SyntaxKind.EndOfFileToken),
    ts.NodeFlags.None,
  );

  const commonImport = ts.factory.createImportDeclaration(
    undefined,
    ts.factory.createImportClause(
      false,
      ts.factory.createIdentifier("* as Common"),
      undefined,
    ),
    ts.factory.createStringLiteral(`./${OpenApiRqFiles.common}`),
    undefined,
  );

  const commonExport = ts.factory.createExportDeclaration(
    undefined,
    false,
    undefined,
    ts.factory.createStringLiteral(`./${OpenApiRqFiles.common}`),
    undefined,
  );

  const queriesExport = ts.factory.createExportDeclaration(
    undefined,
    false,
    undefined,
    ts.factory.createStringLiteral(`./${OpenApiRqFiles.queries}`),
    undefined,
  );

  const mainSource = ts.factory.createSourceFile(
    [commonImport, ...imports, ...exports.mainExports],
    ts.factory.createToken(ts.SyntaxKind.EndOfFileToken),
    ts.NodeFlags.None,
  );

  const suspenseSource = ts.factory.createSourceFile(
    [commonImport, ...imports, ...exports.suspenseExports],
    ts.factory.createToken(ts.SyntaxKind.EndOfFileToken),
    ts.NodeFlags.None,
  );

  const indexSource = ts.factory.createSourceFile(
    [commonExport, queriesExport],
    ts.factory.createToken(ts.SyntaxKind.EndOfFileToken),
    ts.NodeFlags.None,
  );

  const prefetchSource = ts.factory.createSourceFile(
    [commonImport, ...imports, ...exports.allPrefetchExports],
    ts.factory.createToken(ts.SyntaxKind.EndOfFileToken),
    ts.NodeFlags.None,
  );

  return {
    commonSource,
    mainSource,
    suspenseSource,
    indexSource,
    prefetchSource,
  };
};

export const createSource = async ({
  outputPath,
  version,
}: {
  outputPath: string;
  version: string;
}) => {
  const queriesFile = ts.createSourceFile(
    `${OpenApiRqFiles.queries}.ts`,
    "",
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
  const commonFile = ts.createSourceFile(
    `${OpenApiRqFiles.common}.ts`,
    "",
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
  const suspenseFile = ts.createSourceFile(
    `${OpenApiRqFiles.suspense}.ts`,
    "",
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );

  const indexFile = ts.createSourceFile(
    `${OpenApiRqFiles.index}.ts`,
    "",
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );

  const prefetchFile = ts.createSourceFile(
    `${OpenApiRqFiles.prefetch}.ts`,
    "",
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );

  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
    removeComments: false,
  });

  const {
    commonSource,
    mainSource,
    suspenseSource,
    indexSource,
    prefetchSource,
  } = await createSourceFile(outputPath);

  const comment = `// generated with @7nohe/openapi-react-query-codegen@${version} \n\n`;

  const commonResult =
    comment +
    printer.printNode(ts.EmitHint.Unspecified, commonSource, commonFile);

  const mainResult =
    comment +
    printer.printNode(ts.EmitHint.Unspecified, mainSource, queriesFile);

  const suspenseResult =
    comment +
    printer.printNode(ts.EmitHint.Unspecified, suspenseSource, suspenseFile);

  const indexResult =
    comment +
    printer.printNode(ts.EmitHint.Unspecified, indexSource, indexFile);

  const prefetchResult =
    comment +
    printer.printNode(ts.EmitHint.Unspecified, prefetchSource, prefetchFile);

  return [
    {
      name: `${OpenApiRqFiles.index}.ts`,
      content: indexResult,
    },
    {
      name: `${OpenApiRqFiles.common}.ts`,
      content: commonResult,
    },
    {
      name: `${OpenApiRqFiles.queries}.ts`,
      content: mainResult,
    },
    {
      name: `${OpenApiRqFiles.suspense}.ts`,
      content: suspenseResult,
    },
    {
      name: `${OpenApiRqFiles.prefetch}.ts`,
      content: prefetchResult,
    },
  ];
};
