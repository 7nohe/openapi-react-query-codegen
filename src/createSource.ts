import ts from "typescript";
import { createImports } from "./createImports";
import { createExports } from "./createExports";
import { version } from "../package.json";

const createSourceFile = (outputPath: string) => {
  return ts.factory.createSourceFile(
    [...createImports(outputPath), ...createExports(outputPath)],
    ts.factory.createToken(ts.SyntaxKind.EndOfFileToken),
    ts.NodeFlags.None
  );
};

export const createSource = (outputPath: string) => {
  const resultFile = ts.createSourceFile(
    "index.ts",
    "",
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS
  );
  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
    removeComments: false,
  });

  const node = createSourceFile(outputPath);

  const result =
    `// generated with @7nohe/openapi-react-query-codegen@${version} \n` +
    printer.printNode(ts.EmitHint.Unspecified, node, resultFile);

  return result;
};
