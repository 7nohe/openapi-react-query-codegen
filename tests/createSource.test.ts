import { afterAll, beforeAll, describe, expect, test } from "vitest";
import { createSource } from "../src/createSource.mjs";
import { cleanOutputs, generateTSClients, outputPath } from "./utils";
const fileName = "createSource";
describe(fileName, () => {
  beforeAll(async () => await generateTSClients(fileName));
  afterAll(async () => await cleanOutputs(fileName));

  test("createSource", async () => {
    const source = await createSource({
      outputPath: outputPath(fileName),
      version: "1.0.0",
    });

    const indexTs = source.find((s) => s.name === "index.ts");
    expect(indexTs?.content).toMatchSnapshot();

    const commonTs = source.find((s) => s.name === "common.ts");
    expect(commonTs?.content).toMatchSnapshot();

    const queriesTs = source.find((s) => s.name === "queries.ts");
    expect(queriesTs?.content).toMatchSnapshot();

    const suspenseTs = source.find((s) => s.name === "suspense.ts");
    expect(suspenseTs?.content).toMatchSnapshot();

    const prefetchTs = source.find((s) => s.name === "prefetch.ts");
    expect(prefetchTs?.content).toMatchSnapshot();
  });
});
