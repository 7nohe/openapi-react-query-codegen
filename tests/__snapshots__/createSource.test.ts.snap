// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createSource > createSource 1`] = `
"// generated with @7nohe/openapi-react-query-codegen@1.0.0 

export * from "./common";
export * from "./queries";
"
`;

exports[`createSource > createSource 2`] = `
"// generated with @7nohe/openapi-react-query-codegen@1.0.0 

import { type QueryClient, useQuery, useSuspenseQuery, useMutation, UseQueryResult, UseQueryOptions, UseMutationOptions, UseMutationResult } from "@tanstack/react-query";
import { findPets, addPet, getNotDefined, postNotDefined, findPetById, deletePet } from "../requests/services.gen";
import { Pet, NewPet, Error, FindPetsData, FindPetsResponse, FindPetsError, AddPetData, AddPetResponse, AddPetError, GetNotDefinedResponse, GetNotDefinedError, PostNotDefinedResponse, PostNotDefinedError, FindPetByIdData, FindPetByIdResponse, FindPetByIdError, DeletePetData, DeletePetResponse, DeletePetError, $OpenApiTs } from "../requests/types.gen";
export type FindPetsDefaultResponse = Awaited<ReturnType<typeof findPets>>["data"];
export type FindPetsQueryResult<TData = FindPetsDefaultResponse, TError = unknown> = UseQueryResult<TData, TError>;
export const useFindPetsKey = "FindPets";
export const UseFindPetsKeyFn = ({}: {} = {}, queryKey?: Array<unknown>) => [useFindPetsKey, ...(queryKey ?? [{}])];
export type GetNotDefinedDefaultResponse = Awaited<ReturnType<typeof getNotDefined>>["data"];
export type GetNotDefinedQueryResult<TData = GetNotDefinedDefaultResponse, TError = unknown> = UseQueryResult<TData, TError>;
export const useGetNotDefinedKey = "GetNotDefined";
export const UseGetNotDefinedKeyFn = ({}: {} = {}, queryKey?: Array<unknown>) => [useGetNotDefinedKey, ...(queryKey ?? [{}])];
export type FindPetByIdDefaultResponse = Awaited<ReturnType<typeof findPetById>>["data"];
export type FindPetByIdQueryResult<TData = FindPetByIdDefaultResponse, TError = unknown> = UseQueryResult<TData, TError>;
export const useFindPetByIdKey = "FindPetById";
export const UseFindPetByIdKeyFn = ({}: {} = {}, queryKey?: Array<unknown>) => [useFindPetByIdKey, ...(queryKey ?? [{}])];
export type AddPetMutationResult = Awaited<ReturnType<typeof addPet>>;
export type PostNotDefinedMutationResult = Awaited<ReturnType<typeof postNotDefined>>;
export type DeletePetMutationResult = Awaited<ReturnType<typeof deletePet>>;
"
`;

exports[`createSource > createSource 3`] = `
"// generated with @7nohe/openapi-react-query-codegen@1.0.0 

import * as Common from "./common";
import { type QueryClient, useQuery, useSuspenseQuery, useMutation, UseQueryResult, UseQueryOptions, UseMutationOptions, UseMutationResult } from "@tanstack/react-query";
import { findPets, addPet, getNotDefined, postNotDefined, findPetById, deletePet } from "../requests/services.gen";
import { Pet, NewPet, Error, FindPetsData, FindPetsResponse, FindPetsError, AddPetData, AddPetResponse, AddPetError, GetNotDefinedResponse, GetNotDefinedError, PostNotDefinedResponse, PostNotDefinedError, FindPetByIdData, FindPetByIdResponse, FindPetByIdError, DeletePetData, DeletePetResponse, DeletePetError, $OpenApiTs } from "../requests/types.gen";
export const useFindPets = <TData = Common.FindPetsDefaultResponse, TError = unknown, TQueryKey extends Array<unknown> = unknown[]>({}: {} = {}, queryKey?: TQueryKey, options?: Omit<UseQueryOptions<TData, TError>, "queryKey" | "queryFn">) => useQuery<TData, TError>({ queryKey: Common.UseFindPetsKeyFn({}, queryKey), queryFn: () => findPets({}).then(response => response.data as TData) as TData, ...options });
export const useGetNotDefined = <TData = Common.GetNotDefinedDefaultResponse, TError = unknown, TQueryKey extends Array<unknown> = unknown[]>({}: {} = {}, queryKey?: TQueryKey, options?: Omit<UseQueryOptions<TData, TError>, "queryKey" | "queryFn">) => useQuery<TData, TError>({ queryKey: Common.UseGetNotDefinedKeyFn({}, queryKey), queryFn: () => getNotDefined({}).then(response => response.data as TData) as TData, ...options });
export const useFindPetById = <TData = Common.FindPetByIdDefaultResponse, TError = unknown, TQueryKey extends Array<unknown> = unknown[]>({}: {} = {}, queryKey?: TQueryKey, options?: Omit<UseQueryOptions<TData, TError>, "queryKey" | "queryFn">) => useQuery<TData, TError>({ queryKey: Common.UseFindPetByIdKeyFn({}, queryKey), queryFn: () => findPetById({}).then(response => response.data as TData) as TData, ...options });
export const useAddPet = <TData = Common.AddPetMutationResult, TError = unknown, TContext = unknown>(options?: Omit<UseMutationOptions<TData, TError, {}, TContext>, "mutationFn">) => useMutation<TData, TError, {}, TContext>({ mutationFn: ({}) => addPet({}) as unknown as Promise<TData>, ...options });
export const usePostNotDefined = <TData = Common.PostNotDefinedMutationResult, TError = unknown, TContext = unknown>(options?: Omit<UseMutationOptions<TData, TError, {}, TContext>, "mutationFn">) => useMutation<TData, TError, {}, TContext>({ mutationFn: ({}) => postNotDefined({}) as unknown as Promise<TData>, ...options });
export const useDeletePet = <TData = Common.DeletePetMutationResult, TError = unknown, TContext = unknown>(options?: Omit<UseMutationOptions<TData, TError, {}, TContext>, "mutationFn">) => useMutation<TData, TError, {}, TContext>({ mutationFn: ({}) => deletePet({}) as unknown as Promise<TData>, ...options });
"
`;

exports[`createSource > createSource 4`] = `
"// generated with @7nohe/openapi-react-query-codegen@1.0.0 

import * as Common from "./common";
import { type QueryClient, useQuery, useSuspenseQuery, useMutation, UseQueryResult, UseQueryOptions, UseMutationOptions, UseMutationResult } from "@tanstack/react-query";
import { findPets, addPet, getNotDefined, postNotDefined, findPetById, deletePet } from "../requests/services.gen";
import { Pet, NewPet, Error, FindPetsData, FindPetsResponse, FindPetsError, AddPetData, AddPetResponse, AddPetError, GetNotDefinedResponse, GetNotDefinedError, PostNotDefinedResponse, PostNotDefinedError, FindPetByIdData, FindPetByIdResponse, FindPetByIdError, DeletePetData, DeletePetResponse, DeletePetError, $OpenApiTs } from "../requests/types.gen";
export const useFindPetsSuspense = <TData = Common.FindPetsDefaultResponse, TError = unknown, TQueryKey extends Array<unknown> = unknown[]>({}: {} = {}, queryKey?: TQueryKey, options?: Omit<UseQueryOptions<TData, TError>, "queryKey" | "queryFn">) => useSuspenseQuery<TData, TError>({ queryKey: Common.UseFindPetsKeyFn({}, queryKey), queryFn: () => findPets({}).then(response => response.data as TData) as TData, ...options });
export const useGetNotDefinedSuspense = <TData = Common.GetNotDefinedDefaultResponse, TError = unknown, TQueryKey extends Array<unknown> = unknown[]>({}: {} = {}, queryKey?: TQueryKey, options?: Omit<UseQueryOptions<TData, TError>, "queryKey" | "queryFn">) => useSuspenseQuery<TData, TError>({ queryKey: Common.UseGetNotDefinedKeyFn({}, queryKey), queryFn: () => getNotDefined({}).then(response => response.data as TData) as TData, ...options });
export const useFindPetByIdSuspense = <TData = Common.FindPetByIdDefaultResponse, TError = unknown, TQueryKey extends Array<unknown> = unknown[]>({}: {} = {}, queryKey?: TQueryKey, options?: Omit<UseQueryOptions<TData, TError>, "queryKey" | "queryFn">) => useSuspenseQuery<TData, TError>({ queryKey: Common.UseFindPetByIdKeyFn({}, queryKey), queryFn: () => findPetById({}).then(response => response.data as TData) as TData, ...options });
"
`;

exports[`createSource > createSource 5`] = `
"// generated with @7nohe/openapi-react-query-codegen@1.0.0 

import * as Common from "./common";
import { type QueryClient, useQuery, useSuspenseQuery, useMutation, UseQueryResult, UseQueryOptions, UseMutationOptions, UseMutationResult } from "@tanstack/react-query";
import { findPets, addPet, getNotDefined, postNotDefined, findPetById, deletePet } from "../requests/services.gen";
import { Pet, NewPet, Error, FindPetsData, FindPetsResponse, FindPetsError, AddPetData, AddPetResponse, AddPetError, GetNotDefinedResponse, GetNotDefinedError, PostNotDefinedResponse, PostNotDefinedError, FindPetByIdData, FindPetByIdResponse, FindPetByIdError, DeletePetData, DeletePetResponse, DeletePetError, $OpenApiTs } from "../requests/types.gen";
export const prefetchUseFindPets = (queryClient: QueryClient, {}: {} = {}) => queryClient.prefetchQuery({ queryKey: Common.UseFindPetsKeyFn({}), queryFn: () => findPets({}) });
export const prefetchUseGetNotDefined = (queryClient: QueryClient, {}: {} = {}) => queryClient.prefetchQuery({ queryKey: Common.UseGetNotDefinedKeyFn({}), queryFn: () => getNotDefined({}) });
export const prefetchUseFindPetById = (queryClient: QueryClient, {}: {} = {}) => queryClient.prefetchQuery({ queryKey: Common.UseFindPetByIdKeyFn({}), queryFn: () => findPetById({}) });
"
`;
